nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullHandler.java

```diff
@@ -0,0 +1,218 @@
+import javax.lang.model.element.Element;
+import org.checkerframework.dataflow.cfg.node.MethodInvocationNode;
+
+/**
+ * This Handler parses {@code @EnsuresNonNull} annotation and when the annotated method is invoked,
+ * it injects the knowledge gained from the annotation to the data flow analysis. The following
+ * tasks are performed when the {@code @EnsuresNonNull} annotation has observed:
+ *
+ * <ul>
+ *   <li>It validates the syntax of the annotation.
+ *   <li>It validates whether all fields specified in the annotation are guaranteed to be {@code
+ *       Nonnull} at exit point of the method.
+ *   <li>It validates whether the specified postcondition conforms to the overriding rules. Every
+ *       methods postcondition must satisfy all postconditions of the super methods as well.
+ * </ul>
+ */
+public class EnsuresNonNullHandler extends AbstractFieldContractHandler {
+
+  @Override
+  public void onMatchTopLevelClass(
+      NullAway analysis, ClassTree tree, VisitorState state, Symbol.ClassSymbol classSymbol) {
+    annotName = "EnsuresNonNull";
+  }
+
+  /**
+   * Validates whether all parameters mentioned in the @EnsuresNonNull annotation are guaranteed to
+   * be {@code @NonNull} at exit point of this method.
+   */
+  @Override
+  protected boolean validateAnnotationSemantics(
+      NullAway analysis, VisitorState state, MethodTree tree, Symbol.MethodSymbol methodSymbol) {
+    String message;
+    if (tree.getBody() == null) {
+      message = "cannot annotate an abstract method with @EnsuresNonNull annotation";
+
+      state.reportMatch(
+          analysis
+              .getErrorBuilder()
+              .createErrorDescription(
+                  new ErrorMessage(ErrorMessage.MessageTypes.ANNOTATION_VALUE_INVALID, message),
+                  tree,
+                  buildDescriptionFromChecker(tree, analysis),
+                  state));
+      return false;
+    }
+    Set<String> nonnullFieldsOfReceiverAtExit =
+        analysis
+            .getNullnessAnalysis(state)
+            .getNonnullFieldsOfReceiverAtExit(new TreePath(state.getPath(), tree), state.context)
+            .stream()
+            .map(e -> e.getSimpleName().toString())
+            .collect(Collectors.toSet());
+    Set<String> fieldNames = ContractUtils.getFieldNamesFromAnnotation(methodSymbol, annotName);
+    if (fieldNames == null) {
+      fieldNames = Collections.emptySet();
+    }
+    fieldNames = ContractUtils.trimReceivers(fieldNames);
+    boolean isValidLocalPostCondition = nonnullFieldsOfReceiverAtExit.containsAll(fieldNames);
+    if (!isValidLocalPostCondition) {
+      fieldNames.removeAll(nonnullFieldsOfReceiverAtExit);
+      message =
+          "method: "
+              + methodSymbol
+              + " is annotated with @EnsuresNonNull annotation, it indicates that all fields in the annotation parameter"
+              + " must be guaranteed to be nonnull at exit point and it fails to do so for the fields: "
```

Comment:
Agreed it might be good to track an issue on giving a better error here.  I don't think Checker Framework gives a very informative error message for this issue either.  Yes, it does get tricky with multiple exits.